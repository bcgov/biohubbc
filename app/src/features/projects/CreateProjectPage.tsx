import {
  Box,
  Button,
  Container,
  makeStyles,
  Paper,
  Step,
  StepContent,
  StepLabel,
  Stepper,
  Typography
} from '@material-ui/core';
import EditFormControlsComponent from 'components/form/EditFormControlsComponent';
import FormContainer from 'components/form/FormContainer';
import {
  projectAgencyTemplate,
  projectFundingAgencyTemplate,
  projectProponentTemplate,
  projectTemplate
} from 'constants/project-templates';
import React from 'react';
import { useHistory } from 'react-router';

const useStyles = makeStyles((theme) => ({
  stepper: {
    backgroundColor: 'transparent'
  },
  actionsContainer: {
    marginBottom: theme.spacing(2)
  },
  actionButton: {
    marginTop: theme.spacing(1),
    marginRight: theme.spacing(1)
  },
  finishContainer: {
    padding: theme.spacing(3),
    backgroundColor: 'transparent'
  }
}));

/**
 * Page for creating a new project.
 *
 * @return {*}
 */
const CreateProjectPage: React.FC = () => {
  const history = useHistory();

  const classes = useStyles();

  const [activeStep, setActiveStep] = React.useState(0);

  function getSteps() {
    return ['Project Details', 'Project Funding', 'Project Agency', 'Project Proponent'];
  }

  function getStepContent(step: any) {
    switch (step) {
      case 0:
        return <FormContainer template={projectTemplate} formControlsComponent={EditFormControlsComponent} />;
      case 1:
        return (
          <FormContainer template={projectFundingAgencyTemplate} formControlsComponent={EditFormControlsComponent} />
        );
      case 2:
        return <FormContainer template={projectAgencyTemplate} formControlsComponent={EditFormControlsComponent} />;
      case 3:
        return <FormContainer template={projectProponentTemplate} formControlsComponent={EditFormControlsComponent} />;
      default:
        return 'Unknown step';
    }
  }

  const steps = getSteps();

  const navigateToProjectsPage = () => {
    history.push('/projects');
  };

  const handleCancel = () => {
    // TODO are you sure messaging?
    navigateToProjectsPage();
  };

  const handleNext = () => {
    setActiveStep((prevActiveStep) => prevActiveStep + 1);
  };

  const handleBack = () => {
    setActiveStep((prevActiveStep) => prevActiveStep - 1);
  };

  const handleSubmit = () => [];

  return (
    <Box my={3}>
      <Container>
        <Box mb={3}>
          <Typography variant="h2">Create Project</Typography>
        </Box>
        <Box>
          <Stepper activeStep={activeStep} orientation="vertical" className={classes.stepper}>
            {steps.map((label, index) => (
              <Step key={label}>
                <StepLabel>
                  <Typography variant="h3">{label}</Typography>
                </StepLabel>
                <StepContent>
                  {getStepContent(index)}
                  <Box display="flex" justifyContent="space-between" className={classes.actionsContainer}>
                    <Box>
                      <Button variant="contained" onClick={handleCancel} className={classes.actionButton}>
                        Cancel
                      </Button>
                    </Box>
                    <Box>
                      <Button
                        disabled={activeStep === 0}
                        variant="contained"
                        onClick={handleBack}
                        className={classes.actionButton}>
                        Previous
                      </Button>
                      <Button variant="contained" color="primary" onClick={handleNext} className={classes.actionButton}>
                        Next
                      </Button>
                    </Box>
                  </Box>
                </StepContent>
              </Step>
            ))}
          </Stepper>
          {activeStep === steps.length && (
            <Paper square elevation={0} className={classes.finishContainer}>
              <Box mb={3}>
                <Typography variant="h3">All steps complete!</Typography>
              </Box>
              <Box display="flex" justifyContent="space-between" className={classes.actionsContainer}>
                <Box>
                  <Button variant="contained" onClick={handleCancel} className={classes.actionButton}>
                    Cancel
                  </Button>
                </Box>
                <Box>
                  <Button variant="contained" onClick={handleBack} className={classes.actionButton}>
                    Previous
                  </Button>
                  <Button variant="contained" color="primary" onClick={handleSubmit} className={classes.actionButton}>
                    Create Project
                  </Button>
                </Box>
              </Box>
            </Paper>
          )}
        </Box>
      </Container>
    </Box>
  );
};

export default CreateProjectPage;
