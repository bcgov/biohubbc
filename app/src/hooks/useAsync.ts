import { useRef } from 'react';

export type AsyncFunction<Q extends any[], R> = (...args: Q) => Promise<R>;

export type WrappedAsyncFunction<Q extends any[], S> = (...args: Q) => Promise<S>;

/**
 * Wraps an async function to prevent duplicate calls if the previous call is still pending.
 *
 * Example:
 *
 * ```
 * const myAsyncFunction = useAsync(
 *   (param: number) => asyncFunction(param1)
 * )
 *
 * await myAsyncFunction(1) // calls `asyncFunction`, where param=1
 * (call 1 is pending)
 * await myAsyncFunction(2) // returns promise from call 1
 * (call 1 is fulfilled)
 * await myAsyncFunction(2) // calls `asyncFunction`, where param=2
 * (call 2 is pending)
 * ```
 *
 * @template Q
 * @template R
 * @param {AsyncFunction<Q, R>} asyncFunction the async function to wrap
 * @param {IUseAsyncOptions} [options]
 * @return {*}  {WrappedAsyncFunction<Q, R>}
 */
export const useAsync = <Q extends any[], R>(asyncFunction: AsyncFunction<Q, R>): WrappedAsyncFunction<Q, R> => {
  const ref = useRef<Promise<R>>();

  const isPending = useRef(false);

  const wrappedAsyncFunction: WrappedAsyncFunction<Q, R> = async (...args) => {
    if (ref.current && isPending.current) {
      return ref.current;
    }

    isPending.current = true;

    ref.current = asyncFunction(...args)
      .then(
        (response: R) => {
          isPending.current = false;

          return response;
        },
        (error) => {
          isPending.current = false;

          throw error;
        }
      )
      .catch((error) => {
        isPending.current = false;

        throw error;
      });

    return ref.current;
  };

  return wrappedAsyncFunction;
};
