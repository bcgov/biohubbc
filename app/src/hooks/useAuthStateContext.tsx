import { useContext } from 'react';
import { AuthStateContext, IAuthState } from '../contexts/authStateContext';

/**
 * Returns an instance of `IAuthState` from `AuthStateContext`.
 *
 * @return {*}  {IAuthState}
 */
export const useAuthStateContext = (): IAuthState => {
  const context = useContext(AuthStateContext);

  if (!context) {
    throw Error(
      'AuthStateContext is undefined, please verify you are calling useAuthStateContext() as child of an <AuthStateProvider> component.'
    );
  }

  return context;
};
