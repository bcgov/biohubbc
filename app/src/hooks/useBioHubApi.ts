import { useKeycloak } from '@react-keycloak/web';
import axios from 'axios';
import { IProject, IProjectPostObject } from 'interfaces/project-interfaces';
import { IActivity, ICreateActivity, ICreateProjectResponse, ITemplate } from 'interfaces/useBioHubApi-interfaces';
import moment from 'moment';
import { useMemo } from 'react';

const API_HOST = process.env.REACT_APP_API_HOST;
const API_PORT = process.env.REACT_APP_API_PORT;

const API_URL = (API_PORT && `${API_HOST}:${API_PORT}`) || API_HOST || 'api-dev-biohubbc.apps.silver.devops.gov.bc.ca';

/**
 * Returns an instance of axios with baseURL and authorization headers set.
 *
 * @return {*}
 */
const useApi = () => {
  const { keycloak } = useKeycloak();
  const instance = useMemo(() => {
    return axios.create({
      headers: {
        Authorization: `Bearer ${keycloak?.token}`
      },
      baseURL: API_URL
    });
  }, [keycloak]);

  return instance;
};

/**
 * Returns a set of supported api methods.
 *
 * @return {object} object whose properties are supported api methods.
 */
export const useBiohubApi = () => {
  const api = useApi();

  /**
   * Get all projects.
   *
   * @return {*}  {Promise<IProject[]>}
   */
  const getProjects = async (): Promise<IProject[]> => {
    // const { data } = await api.get(`/api/projects`);

    // return data;

    // TODO: stub for development
    return [
      {
        id: 1,
        name: 'Project Name 1',
        objectives: 'Project Objectives 1',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 1',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 1',
        caveats: 'Caveats 1',
        comments: 'Comments 1'
      },
      {
        id: 2,
        name: 'Project Name 2',
        objectives: 'Project Objectives 2',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 2',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 2',
        caveats: 'Caveats 2',
        comments: 'Comments 2'
      },
      {
        id: 3,
        name: 'Project Name 3',
        objectives: 'Project Objectives 3',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 3',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 3',
        caveats: 'Caveats 3',
        comments: 'Comments 3'
      },
      {
        id: 4,
        name: 'Project Name 4',
        objectives: 'Project Objectives 4',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 4',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 4',
        caveats: 'Caveats 4',
        comments: 'Comments 4'
      }
    ];
  };

  /**
   * Get a project based on its ID.
   *
   * @param {projectId} projectId
   * @return {*}  {Promise<IProject>}
   */
  const getProject = async (projectId: number): Promise<IProject> => {
    const { data } = await api.get(`/api/project/${projectId}`);

    return data;
  };

  /**
   * Create a new project.
   *
   * @param {IProjectPostObject} project
   * @return {*}  {Promise<ICreateProjectResponse>}
   */
  const createProject = async (project: IProjectPostObject): Promise<ICreateProjectResponse> => {
    const { data } = await api.post('/api/project', project);

    return data;
  };

  /**
   * Get a template based on its ID.
   *
   * @param {templateId} templateId
   * @return {*}  {Promise<ITemplate>}
   */
  const getTemplate = async (templateId: number): Promise<ITemplate> => {
    const { data } = await api.get(`/api/template/${templateId}`);

    return data;
  };

  /**
   * Create a new activity record.
   *
   * @param {ICreateActivity} activity
   * @return {*}  {Promise<IActivity>}
   */
  const createActivity = async (activity: ICreateActivity): Promise<IActivity> => {
    const { data } = await api.post('/api/activity', activity);

    return data;
  };

  /**
   * Fetch all code sets.
   *
   * @return {*}  {Promise<any>}
   */
  const getAllCodes = async (): Promise<any> => {
    const { data } = await api.get('/api/codes/');

    return data;
  };

  /**
   * Fetch the api json-schema spec.
   *
   * @return {*}  {Promise<any>}
   */
  const getApiSpec = async (): Promise<any> => {
    const { data } = await api.get('/api/api-docs/');

    return data;
  };

  return {
    getProjects,
    getProject,
    getTemplate,
    createProject,
    createActivity,
    getAllCodes,
    getApiSpec
  };
};
