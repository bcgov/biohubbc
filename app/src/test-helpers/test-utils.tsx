import { ThemeProvider } from '@mui/material/styles';
import { render, RenderOptions } from '@testing-library/react';
import React, { PropsWithChildren, ReactElement } from 'react';
import appTheme from 'themes/appTheme';

const AllProviders = (props: PropsWithChildren<Record<never, unknown>>) => {
  const { children } = props;
  return <ThemeProvider theme={appTheme}>{children}</ThemeProvider>;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
