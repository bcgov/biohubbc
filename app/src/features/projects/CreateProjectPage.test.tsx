import { cleanup, findByText as rawFindByText, fireEvent, render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { useBiohubApi } from 'hooks/useBioHubApi';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { Router } from 'react-router';
import CreateProjectPage from './CreateProjectPage';

const history = createMemoryHistory();

jest.mock('../../hooks/useBioHubApi');
const mockUseBiohubApi = {
  getAllCodes: jest.fn<Promise<object>, []>()
};

const mockBiohubApi = ((useBiohubApi as unknown) as jest.Mock<typeof mockUseBiohubApi>).mockReturnValue(
  mockUseBiohubApi
);

const renderContainer = () => {
  return render(
    <Router history={history}>
      <CreateProjectPage />,
    </Router>
  );
};

describe('CreateProjectPage', () => {
  beforeEach(() => {
    // clear mocks before each test
    mockBiohubApi().getAllCodes.mockClear();
  });

  afterEach(() => {
    cleanup();
  });

  it('renders the initial default page correctly', async () => {
    await act(async () => {
      mockBiohubApi().getAllCodes.mockResolvedValue({
        code_set: []
      });
      const { findByText, asFragment } = renderContainer();

      const projectCorrdinatorHeader = await findByText('Project Coordinator', { selector: 'h2' });
      expect(projectCorrdinatorHeader).toBeVisible();

      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('shows the page title', async () => {
    await act(async () => {
      mockBiohubApi().getAllCodes.mockResolvedValue({
        code_set: []
      });
      const { findByText } = renderContainer();
      const PageTitle = await findByText('Create Project');

      expect(PageTitle).toBeVisible();
    });
  });

  describe('Are you sure? Dialog', () => {
    it('shows warning dialog if the user clicks the `Cancel and Exit` button', async () => {
      await act(async () => {
        mockBiohubApi().getAllCodes.mockResolvedValue({
          code_set: []
        });
        history.push('/home');
        history.push('/projects/create');
        const { findByText, getByRole } = renderContainer();
        const BackToProjectsButton = await findByText('Cancel and Exit', { exact: false });

        fireEvent.click(BackToProjectsButton);
        const AreYouSureTitle = await findByText('Cancel Create Project');
        const AreYouSureText = await findByText('Are you sure you want to cancel?');
        const AreYouSureYesButton = await rawFindByText(getByRole('dialog'), 'Yes', { exact: false });

        expect(AreYouSureTitle).toBeVisible();
        expect(AreYouSureText).toBeVisible();
        expect(AreYouSureYesButton).toBeVisible();
      });
    });

    it('it calls history.push() if the user clicks `Yes`', async () => {
      await act(async () => {
        mockBiohubApi().getAllCodes.mockResolvedValue({
          code_set: []
        });
        history.push('/home');
        history.push('/projects/create');
        const { findByText, getByRole } = renderContainer();
        const BackToProjectsButton = await findByText('Cancel and Exit', { exact: false });

        fireEvent.click(BackToProjectsButton);
        const AreYouSureYesButton = await rawFindByText(getByRole('dialog'), 'Yes', { exact: false });

        expect(history.location.pathname).toEqual('/projects/create');
        fireEvent.click(AreYouSureYesButton);
        expect(history.location.pathname).toEqual('/projects');
      });
    });

    it('it does nothing if the user clicks `No`', async () => {
      await act(async () => {
        mockBiohubApi().getAllCodes.mockResolvedValue({
          code_set: []
        });
        history.push('/home');
        history.push('/projects/create');
        const { findByText, getByRole } = renderContainer();
        const BackToProjectsButton = await findByText('Cancel and Exit', { exact: false });

        fireEvent.click(BackToProjectsButton);
        const AreYouSureNoButton = await rawFindByText(getByRole('dialog'), 'No', { exact: false });

        expect(history.location.pathname).toEqual('/projects/create');
        fireEvent.click(AreYouSureNoButton);
        expect(history.location.pathname).toEqual('/projects/create');
      });
    });
  });
});
