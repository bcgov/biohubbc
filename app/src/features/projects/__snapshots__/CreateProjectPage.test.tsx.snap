// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateProjectPage renders the initial default page correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-5"
  />
  <div
    class="MuiBox-root MuiBox-root-6"
  />
  <div
    class="MuiBox-root MuiBox-root-7"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg"
    >
      <div
        class="MuiBox-root MuiBox-root-8"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                />
              </svg>
            </span>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Back to Projects
            </p>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root MuiBox-root-9"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2"
        >
          Create Project
        </h2>
      </div>
      <div
        class="MuiBox-root MuiBox-root-10"
      >
        <div
          class="MuiPaper-root MuiStepper-root MuiStepper-vertical makeStyles-stepper-1 MuiPaper-elevation0"
        >
          <div
            class="MuiStep-root MuiStep-vertical"
          >
            <span
              class="MuiStepLabel-root MuiStepLabel-vertical"
            >
              <span
                class="MuiStepLabel-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiStepIcon-root MuiStepIcon-active"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="12"
                  />
                  <text
                    class="MuiStepIcon-text"
                    text-anchor="middle"
                    x="12"
                    y="16"
                  >
                    1
                  </text>
                </svg>
              </span>
              <span
                class="MuiStepLabel-labelContainer"
              >
                <span
                  class="MuiTypography-root MuiStepLabel-label MuiStepLabel-active MuiTypography-body2 MuiTypography-displayBlock"
                >
                  <h2
                    class="MuiTypography-root MuiTypography-h2"
                  >
                    Project Coordinator
                  </h2>
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2"
                  >
                    Enter contact details for the project coordinator
                  </h6>
                </span>
              </span>
            </span>
            <div
              class="MuiStepContent-root"
              icon="1"
            >
              <div
                class="MuiCollapse-container MuiStepContent-transition MuiCollapse-entered"
                style="min-height: 0px;"
              >
                <div
                  class="MuiCollapse-wrapper"
                >
                  <div
                    class="MuiCollapse-wrapperInner"
                  >
                    <form>
                      <div
                        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
                      >
                        <div
                          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
                        >
                          <div
                            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                          >
                            <label
                              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                              data-shrink="false"
                              for="first_name"
                              id="first_name-label"
                            >
                              First Name
                              <span
                                aria-hidden="true"
                                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                              >
                                 *
                              </span>
                            </label>
                            <div
                              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                            >
                              <input
                                aria-invalid="false"
                                class="MuiInputBase-input MuiOutlinedInput-input"
                                id="first_name"
                                name="first_name"
                                required=""
                                type="text"
                                value=""
                              />
                              <fieldset
                                aria-hidden="true"
                                class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
                              >
                                <legend
                                  class="PrivateNotchedOutline-legendLabelled-13"
                                >
                                  <span>
                                    First Name *
                                  </span>
                                </legend>
                              </fieldset>
                            </div>
                          </div>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
                        >
                          <div
                            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                          >
                            <label
                              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                              data-shrink="false"
                              for="last_name"
                              id="last_name-label"
                            >
                              Last Name
                              <span
                                aria-hidden="true"
                                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                              >
                                 *
                              </span>
                            </label>
                            <div
                              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                            >
                              <input
                                aria-invalid="false"
                                class="MuiInputBase-input MuiOutlinedInput-input"
                                id="last_name"
                                name="last_name"
                                required=""
                                type="text"
                                value=""
                              />
                              <fieldset
                                aria-hidden="true"
                                class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
                              >
                                <legend
                                  class="PrivateNotchedOutline-legendLabelled-13"
                                >
                                  <span>
                                    Last Name *
                                  </span>
                                </legend>
                              </fieldset>
                            </div>
                          </div>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                        >
                          <div
                            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                          >
                            <label
                              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                              data-shrink="false"
                              for="email_address"
                              id="email_address-label"
                            >
                              Business Email Address
                              <span
                                aria-hidden="true"
                                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                              >
                                 *
                              </span>
                            </label>
                            <div
                              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
                            >
                              <input
                                aria-invalid="false"
                                class="MuiInputBase-input MuiOutlinedInput-input"
                                id="email_address"
                                name="email_address"
                                required=""
                                type="text"
                                value=""
                              />
                              <fieldset
                                aria-hidden="true"
                                class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
                              >
                                <legend
                                  class="PrivateNotchedOutline-legendLabelled-13"
                                >
                                  <span>
                                    Business Email Address *
                                  </span>
                                </legend>
                              </fieldset>
                            </div>
                          </div>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                        >
                          <div
                            aria-expanded="false"
                            class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon"
                            role="combobox"
                          >
                            <div
                              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                            >
                              <label
                                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
                                data-shrink="false"
                                for="coordinator_agency"
                                id="coordinator_agency-label"
                              >
                                Coordinator Agency
                                <span
                                  aria-hidden="true"
                                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                                >
                                   *
                                </span>
                              </label>
                              <div
                                class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
                              >
                                <input
                                  aria-autocomplete="list"
                                  aria-invalid="false"
                                  autocapitalize="none"
                                  autocomplete="off"
                                  class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                                  id="coordinator_agency"
                                  required=""
                                  spellcheck="false"
                                  type="text"
                                  value=""
                                />
                                <div
                                  class="MuiAutocomplete-endAdornment"
                                >
                                  <button
                                    aria-label="Clear"
                                    class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator MuiAutocomplete-clearIndicatorDirty"
                                    tabindex="-1"
                                    title="Clear"
                                    type="button"
                                  >
                                    <span
                                      class="MuiIconButton-label"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                                        focusable="false"
                                        viewBox="0 0 24 24"
                                      >
                                        <path
                                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                                        />
                                      </svg>
                                    </span>
                                    <span
                                      class="MuiTouchRipple-root"
                                    />
                                  </button>
                                </div>
                                <fieldset
                                  aria-hidden="true"
                                  class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
                                >
                                  <legend
                                    class="PrivateNotchedOutline-legendLabelled-13"
                                  >
                                    <span>
                                      Coordinator Agency *
                                    </span>
                                  </legend>
                                </fieldset>
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
                        >
                          <fieldset
                            class="MuiFormControl-root"
                          >
                            <legend
                              class="MuiFormLabel-root Mui-required"
                            >
                              Share Contact Details
                              <span
                                aria-hidden="true"
                                class="MuiFormLabel-asterisk"
                              >
                                 *
                              </span>
                            </legend>
                            <div
                              aria-label="Share Contact Details"
                              class="MuiFormGroup-root"
                              role="radiogroup"
                            >
                              <label
                                class="MuiFormControlLabel-root"
                              >
                                <span
                                  aria-disabled="false"
                                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-15 MuiRadio-root MuiRadio-colorSecondary PrivateSwitchBase-checked-16 Mui-checked MuiIconButton-colorSecondary"
                                >
                                  <span
                                    class="MuiIconButton-label"
                                  >
                                    <input
                                      checked=""
                                      class="PrivateSwitchBase-input-18"
                                      name="share_contact_details"
                                      required=""
                                      type="radio"
                                      value="false"
                                    />
                                    <div
                                      class="PrivateRadioButtonIcon-root-19 PrivateRadioButtonIcon-checked-21"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        class="MuiSvgIcon-root"
                                        focusable="false"
                                        viewBox="0 0 24 24"
                                      >
                                        <path
                                          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                                        />
                                      </svg>
                                      <svg
                                        aria-hidden="true"
                                        class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-20"
                                        focusable="false"
                                        viewBox="0 0 24 24"
                                      >
                                        <path
                                          d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                                        />
                                      </svg>
                                    </div>
                                  </span>
                                  <span
                                    class="MuiTouchRipple-root"
                                  />
                                </span>
                                <span
                                  class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
                                >
                                  No
                                </span>
                              </label>
                              <label
                                class="MuiFormControlLabel-root"
                              >
                                <span
                                  aria-disabled="false"
                                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-15 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
                                >
                                  <span
                                    class="MuiIconButton-label"
                                  >
                                    <input
                                      class="PrivateSwitchBase-input-18"
                                      name="share_contact_details"
                                      required=""
                                      type="radio"
                                      value="true"
                                    />
                                    <div
                                      class="PrivateRadioButtonIcon-root-19"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        class="MuiSvgIcon-root"
                                        focusable="false"
                                        viewBox="0 0 24 24"
                                      >
                                        <path
                                          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                                        />
                                      </svg>
                                      <svg
                                        aria-hidden="true"
                                        class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-20"
                                        focusable="false"
                                        viewBox="0 0 24 24"
                                      >
                                        <path
                                          d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                                        />
                                      </svg>
                                    </div>
                                  </span>
                                  <span
                                    class="MuiTouchRipple-root"
                                  />
                                </span>
                                <span
                                  class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
                                >
                                  Yes
                                </span>
                              </label>
                              <p
                                class="MuiFormHelperText-root Mui-required"
                              />
                            </div>
                          </fieldset>
                        </div>
                      </div>
                    </form>
                    <div
                      class="MuiBox-root MuiBox-root-22"
                    >
                      <hr
                        class="MuiDivider-root"
                      />
                    </div>
                    <div
                      class="MuiBox-root MuiBox-root-23 makeStyles-actionsContainer-2"
                    >
                      <div
                        class="MuiBox-root MuiBox-root-24"
                      >
                        <button
                          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-actionButton-3"
                          tabindex="0"
                          type="button"
                        >
                          <span
                            class="MuiButton-label"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1"
                            >
                              Cancel
                            </p>
                          </span>
                          <span
                            class="MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                      <div
                        class="MuiBox-root MuiBox-root-25"
                      >
                        <button
                          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-actionButton-3 Mui-disabled Mui-disabled"
                          disabled=""
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            class="MuiButton-label"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1"
                            >
                              Previous
                            </p>
                          </span>
                        </button>
                        <button
                          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-actionButton-3 MuiButton-containedPrimary"
                          tabindex="0"
                          type="submit"
                        >
                          <span
                            class="MuiButton-label"
                          >
                            <p
                              class="MuiTypography-root MuiTypography-body1"
                            >
                              Next
                            </p>
                          </span>
                          <span
                            class="MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiStepConnector-root MuiStepConnector-vertical Mui-disabled"
          >
            <span
              class="MuiStepConnector-line MuiStepConnector-lineVertical"
            />
          </div>
          <div
            class="MuiStep-root MuiStep-vertical"
          >
            <span
              class="MuiStepLabel-root MuiStepLabel-vertical Mui-disabled"
            >
              <span
                class="MuiStepLabel-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiStepIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="12"
                  />
                  <text
                    class="MuiStepIcon-text"
                    text-anchor="middle"
                    x="12"
                    y="16"
                  >
                    2
                  </text>
                </svg>
              </span>
              <span
                class="MuiStepLabel-labelContainer"
              >
                <span
                  class="MuiTypography-root MuiStepLabel-label MuiTypography-body2 MuiTypography-displayBlock"
                >
                  <h2
                    class="MuiTypography-root MuiTypography-h2"
                  >
                    Permits
                  </h2>
                  <h6
                    class="MuiTypography-root MuiTypography-subtitle2"
                  >
                    Enter permits associated with this project
                  </h6>
                </span>
              </span>
            </span>
            <div
              class="MuiStepContent-root MuiStepContent-last"
              disabled=""
              icon="2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  ,
</DocumentFragment>
`;
